{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import matplotlib as plt\n",
    "from tqdm import tqdm\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Database Design\n",
    "\n",
    "For this assignment, we’ll focus on three pathways: **glycolysis, the citric acid cycle, and the pentose phosphate pathway**. We will consider genes in these pathways from **Drosophila, E. coli, and humans**.\n",
    "\n",
    " - *Gene Table*: name, description, organism, and nucleotide sequence. Additional fields might include chromosome, start and end position, strand, and translated sequence. For eukaryotes, the nucleotide sequence should be the spliced mRNA and the coordinates should span the entire locus.\n",
    "\n",
    " - *Pathway Table*: name and description...\n",
    "\n",
    " - *Enzyme Table:* name, function, and enzyme commission (EC) number. Multiple genes encode enzymes that perform the same function, so there ought to be fewer enzymes than genes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Declaration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect('lab4.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "for tbl in ('genes', 'pathways', 'enzymes'):\n",
    "    c.execute('DROP TABLE IF EXISTS %s;' % tbl);\n",
    "c.execute('''CREATE TABLE genes (\n",
    "    id TEXT,\n",
    "    short TEXT,\n",
    "    long TEXT,\n",
    "    position TEXT,\n",
    "    organism TEXT\n",
    ");''')\n",
    "c.execute('''CREATE TABLE pathways (\n",
    "    keggMap TEXT,\n",
    "    amiGO TEXT,\n",
    "    name TEXT,\n",
    "    description TEXT\n",
    ");''')\n",
    "c.execute('''CREATE TABLE enzymes (\n",
    "    ec TEXT,\n",
    "    name TEXT,\n",
    "    description TEXT,\n",
    "    cas TEXT\n",
    ");''')\n",
    "\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Accessing databases\n",
    "\n",
    "Map of glycolysis from KEGG: https://www.genome.jp/kegg-bin/show_pathway?map00010\n",
    "\n",
    "Databases\n",
    " - UCSC: https://genome.ucsc.edu/\n",
    " - Entrez: https://www.ncbi.nlm.nih.gov/search/\n",
    " - Reactome: https://reactome.org/\n",
    " - Pfam: https://pfam.xfam.org/\n",
    " - KEGG: https://www.genome.jp/kegg/\n",
    "\n",
    "Pick 4 enzymes from glycolysis, TCA cycle, and pentose phosphate. Gather information and sequences from Drosophila, E. coli, and human for those enzymes.\n",
    "\n",
    "Do you have any preference? If you’re clever, you can automate this entire process using Bio.Entrez, which can be used to pull entire annotations for each enzyme from each organism—and more. Try asking it to return data as GenBank and using `Bio.SeqIO` to read the returned GenBank file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataset building\n",
    "\n",
    "You’ll be creating 4 enzymes * 3 pathways * 3 organisms = 36 rows in your gene table. In your pathways table, you should have a row per pathway (3 in total). In your enzymes table, you should have one row per enzyme (12 in total).\n",
    "\n",
    "For this lab, I have decided to pull soley from KEGG using the utilities already in BioPython. Given an ID, it is possible to retrieve enzyme and gene features, but not pathways. So, I start with the manually retrieved information about the pathways and a sample of enzyme ID's associated with each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keggMap amiGO name description\n",
    "pathways = [\n",
    "    # https://www.genome.jp/dbget-bin/www_bget?map00020\n",
    "    ('map00020', '0006099', 'Citrate cycle (TCA cycle)',\n",
    "     'The citrate cycle (TCA cycle, Krebs cycle) is an important aerobic pathway for the final steps of the oxidation of carbohydrates and fatty acids.'),\n",
    "    # https://www.genome.jp/dbget-bin/www_bget?map00030\n",
    "    ('map00030', '0006098', 'Pentose phosphate pathway',\n",
    "     'The pentose phosphate pathway is a process of glucose turnover that produces NADPH as reducing equivalents and pentoses as essential parts of nucleotides.'),\n",
    "    # https://www.genome.jp/dbget-bin/www_bget?map00010\n",
    "    ('map00010', '0006096', 'Glycolysis / Gluconeogenesis',\n",
    "     'Glycolysis is the process of converting glucose into pyruvate and generating small amounts of ATP (energy) and NADH (reducing power).'),\n",
    "]\n",
    "# you can try REST.kegg_get(KEGGmap), however I didn't have enough time to write a parser\n",
    "\n",
    "#txid\n",
    "organisms = [\n",
    "    # https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&name=Homo+sapiens&lvl=0&srchmode=1&keep=1&unlock\n",
    "    ('hsa', 9606, 'Homo sapiens (human)'),\n",
    "    # https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&name=Escherichia+coli&lvl=0&srchmode=1&keep=1&unlock\n",
    "    ('eco', 562, 'Escherichia coli K-12 MG1655'),\n",
    "    # https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=7215\n",
    "    ('dsa', 7215, 'Drosophila melanogaster (fruit fly)')\n",
    "]\n",
    "\n",
    "# https://www.genome.jp/dbget-bin/get_linkdb?-t+enzyme+path:map00020\n",
    "# https://www.genome.jp/dbget-bin/get_linkdb?-t+enzyme+path:map00030\n",
    "# https://www.genome.jp/dbget-bin/get_linkdb?-t+enzyme+path:map00010\n",
    "ecs = [\n",
    "    \"2.7.1.11\",\n",
    "    \"6.2.1.5\",\n",
    "    \"1.8.1.4\",\n",
    "    \"5.1.3.1\",\n",
    "    \"4.1.2.4\",\n",
    "    \"5.4.2.2\",\n",
    "    \"4.2.1.3\",\n",
    "    \"1.2.4.1\",\n",
    "    \"1.1.1.44\",\n",
    "    \"5.3.1.9\",\n",
    "    \"4.2.1.11\",\n",
    "    \"2.7.1.2\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.KEGG import REST, Enzyme, Gene\n",
    "\n",
    "def fetch_enzymes(ecs):\n",
    "    '''Accesses a list of enzyme information from KEGG given EC numbers.'''\n",
    "    records = []\n",
    "    for i in tqdm(range(0, len(ecs), 10)):\n",
    "        handles = REST.kegg_get([\"ec:\"+ec for ec in ecs[i:i+10]])\n",
    "        records.extend(list(Enzyme.parse(handles)))\n",
    "    return records\n",
    "\n",
    "def fetch_genes(ids):\n",
    "    '''Accesses gene information from KEGG given org:id format.'''\n",
    "    records = []\n",
    "    for i in tqdm(range(0, len(ids), 10)):\n",
    "        handles = REST.kegg_get(ids[i:i+10])\n",
    "        records.extend(list(Gene.parse(handles)))\n",
    "    return list(records)\n",
    "\n",
    "def entrez_lookup(query, limit=1):\n",
    "    '''Looks up a query using Entrez.'''\n",
    "    handle = Entrez.esearch(db = 'nucleotide',\n",
    "                            term = 'query',\n",
    "                            sort = 'relevance',\n",
    "                            idtype = 'acc',\n",
    "                            retmax = limit)\n",
    "        \n",
    "    for entry in Entrez.read(handle)['IdList']:  \n",
    "        gb = Entrez.efetch(db='nucleotide', id=entry, rettype='gb', retmode = 'text')\n",
    "        record = SeqIO.read(gb, 'gb')\n",
    "        print(record.name, record.description, record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:02<00:00,  2.03s/it]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['entry', 'name', 'classname', 'sysname', 'reaction', 'substrate', 'product', 'inhibitor', 'cofactor', 'effector', 'comment', 'pathway', 'genes', 'disease', 'structures', 'dblinks'])"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# https://biopython.org/docs/1.75/api/Bio.KEGG.Enzyme.html\n",
    "r = fetch_enzymes(['5.4.2.2'])\n",
    "r[0].__dict__.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  2.32it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['entry', 'name', 'definition', 'orthology', 'organism', 'position', 'motif', 'dblinks'])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# https://biopython.org/docs/1.75/api/Bio.KEGG.Gene.html\n",
    "r = fetch_genes(['hsa:55276'])\n",
    "r[0].__dict__.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shorten_description(text):\n",
    "    '''Keeps the first two sentences of `text`'''\n",
    "    return '. '.join(text.split('. ')[:2]).rstrip('.')+'.'\n",
    "\n",
    "enzymes = []\n",
    "# ec id\n",
    "enz_genes = []\n",
    "# ec map\n",
    "enz_path = []\n",
    "for r in fetch_enzymes(ecs):\n",
    "    enzymes.append((\n",
    "        r.entry,\n",
    "        r.name[0],\n",
    "        shorten_description(r.comment[0]) if len(r.comment) != 0 else '',\n",
    "        r.dblinks[-1][1][0],\n",
    "    ))\n",
    "    genes = dict(r.genes)\n",
    "    for org in ('HSA', 'DME', 'ECO'):\n",
    "        for gene in genes.get(org, []):\n",
    "            enz_genes.append((r.entry, f'{org.lower()}:{gene}'))\n",
    "    for p in r.pathway:\n",
    "        if p[1][-5:] in {'00010', '00020', '00030'}:\n",
    "            enz_path.append((r.entry, 'map'+p[1][-5:]))\n",
    "enzymes, enz_genes, enz_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = []\n",
    "for r in fetch_genes([id for ec, id in enz_genes]):\n",
    "    genes.append((\n",
    "        r.entry,\n",
    "        r.name[0] if len(r.name) != 0 else None,\n",
    "        r.definition,\n",
    "        r.position,\n",
    "        r.organism[0]\n",
    "    ))\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = conn.cursor()\n",
    "c.executemany('INSERT INTO genes VALUES (?,?,?,?,?);', genes)\n",
    "c.executemany('INSERT INTO pathways VALUES (?,?,?,?);', pathways)\n",
    "c.executemany('INSERT INTO enzymes VALUES (?,?,?,?);', enzymes)\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keggMap</th>\n",
       "      <th>amiGO</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>map00020</td>\n",
       "      <td>0006099</td>\n",
       "      <td>Citrate cycle (TCA cycle)</td>\n",
       "      <td>The citrate cycle (TCA cycle, Krebs cycle) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>map00030</td>\n",
       "      <td>0006098</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>The pentose phosphate pathway is a process of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>map00010</td>\n",
       "      <td>0006096</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Glycolysis is the process of converting glucos...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    keggMap    amiGO                          name  \\\n",
       "0  map00020  0006099     Citrate cycle (TCA cycle)   \n",
       "1  map00030  0006098     Pentose phosphate pathway   \n",
       "2  map00010  0006096  Glycolysis / Gluconeogenesis   \n",
       "\n",
       "                                         description  \n",
       "0  The citrate cycle (TCA cycle, Krebs cycle) is ...  \n",
       "1  The pentose phosphate pathway is a process of ...  \n",
       "2  Glycolysis is the process of converting glucos...  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM pathways', conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ec</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>cas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>6-phosphofructokinase</td>\n",
       "      <td>D-Tagatose 6-phosphate and sedoheptulose 7-pho...</td>\n",
       "      <td>9001-80-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6.2.1.5</td>\n",
       "      <td>succinate---CoA ligase (ADP-forming)</td>\n",
       "      <td></td>\n",
       "      <td>9080-33-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.8.1.4</td>\n",
       "      <td>dihydrolipoyl dehydrogenase</td>\n",
       "      <td>A flavoprotein (FAD). A component of the multi...</td>\n",
       "      <td>9001-18-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.1.3.1</td>\n",
       "      <td>ribulose-phosphate 3-epimerase</td>\n",
       "      <td>The enzyme also converts D-erythrose 4-phospha...</td>\n",
       "      <td>9024-20-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.1.2.4</td>\n",
       "      <td>deoxyribose-phosphate aldolase</td>\n",
       "      <td></td>\n",
       "      <td>9026-97-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5.4.2.2</td>\n",
       "      <td>phosphoglucomutase (alpha-D-glucose-1,6-bispho...</td>\n",
       "      <td>Maximum activity is only obtained in the prese...</td>\n",
       "      <td>9001-81-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4.2.1.3</td>\n",
       "      <td>aconitate hydratase</td>\n",
       "      <td>Besides interconverting citrate and cis-aconit...</td>\n",
       "      <td>9024-25-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.2.4.1</td>\n",
       "      <td>pyruvate dehydrogenase (acetyl-transferring)</td>\n",
       "      <td>Contains thiamine diphosphate. It is a compone...</td>\n",
       "      <td>9014-20-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.1.1.44</td>\n",
       "      <td>phosphogluconate dehydrogenase (NADP+-dependen...</td>\n",
       "      <td>The enzyme participates in the oxidative branc...</td>\n",
       "      <td>9073-95-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5.3.1.9</td>\n",
       "      <td>glucose-6-phosphate isomerase</td>\n",
       "      <td>Also catalyses the anomerization of D-glucose ...</td>\n",
       "      <td>9001-41-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4.2.1.11</td>\n",
       "      <td>phosphopyruvate hydratase</td>\n",
       "      <td>Also acts on 3-phospho-D-erythronate.</td>\n",
       "      <td>9014-08-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2.7.1.2</td>\n",
       "      <td>glucokinase</td>\n",
       "      <td>A group of enzymes found in invertebrates and ...</td>\n",
       "      <td>9001-36-9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ec                                               name  \\\n",
       "0   2.7.1.11                              6-phosphofructokinase   \n",
       "1    6.2.1.5               succinate---CoA ligase (ADP-forming)   \n",
       "2    1.8.1.4                        dihydrolipoyl dehydrogenase   \n",
       "3    5.1.3.1                     ribulose-phosphate 3-epimerase   \n",
       "4    4.1.2.4                     deoxyribose-phosphate aldolase   \n",
       "5    5.4.2.2  phosphoglucomutase (alpha-D-glucose-1,6-bispho...   \n",
       "6    4.2.1.3                                aconitate hydratase   \n",
       "7    1.2.4.1       pyruvate dehydrogenase (acetyl-transferring)   \n",
       "8   1.1.1.44  phosphogluconate dehydrogenase (NADP+-dependen...   \n",
       "9    5.3.1.9                      glucose-6-phosphate isomerase   \n",
       "10  4.2.1.11                          phosphopyruvate hydratase   \n",
       "11   2.7.1.2                                        glucokinase   \n",
       "\n",
       "                                          description        cas  \n",
       "0   D-Tagatose 6-phosphate and sedoheptulose 7-pho...  9001-80-3  \n",
       "1                                                      9080-33-5  \n",
       "2   A flavoprotein (FAD). A component of the multi...  9001-18-7  \n",
       "3   The enzyme also converts D-erythrose 4-phospha...  9024-20-8  \n",
       "4                                                      9026-97-5  \n",
       "5   Maximum activity is only obtained in the prese...  9001-81-4  \n",
       "6   Besides interconverting citrate and cis-aconit...  9024-25-3  \n",
       "7   Contains thiamine diphosphate. It is a compone...  9014-20-4  \n",
       "8   The enzyme participates in the oxidative branc...  9073-95-4  \n",
       "9   Also catalyses the anomerization of D-glucose ...  9001-41-6  \n",
       "10              Also acts on 3-phospho-D-erythronate.  9014-08-8  \n",
       "11  A group of enzymes found in invertebrates and ...  9001-36-9  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM enzymes', conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>short</th>\n",
       "      <th>long</th>\n",
       "      <th>position</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5211</td>\n",
       "      <td>PFKL, ATP-PFK, PFK-B, PFK-L</td>\n",
       "      <td>(RefSeq) phosphofructokinase, liver type</td>\n",
       "      <td>21q22.3</td>\n",
       "      <td>hsa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5213</td>\n",
       "      <td>PFKM, ATP-PFK, GSD7, PFK-1, PFK-A, PFK1, PFKA,...</td>\n",
       "      <td>(RefSeq) phosphofructokinase, muscle</td>\n",
       "      <td>12q13.11</td>\n",
       "      <td>hsa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5214</td>\n",
       "      <td>PFKP, ATP-PFK, PFK-C, PFK-P, PFKF</td>\n",
       "      <td>(RefSeq) phosphofructokinase, platelet</td>\n",
       "      <td>10p15.2</td>\n",
       "      <td>hsa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Dmel_CG4001</td>\n",
       "      <td>Pfk</td>\n",
       "      <td>(RefSeq) phosphofructokinase, isoform B</td>\n",
       "      <td>2R</td>\n",
       "      <td>dme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b1723</td>\n",
       "      <td>pfkB</td>\n",
       "      <td>(RefSeq) 6-phosphofructokinase II</td>\n",
       "      <td>1806370..1807299</td>\n",
       "      <td>eco</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>387712</td>\n",
       "      <td>ENO4, C10orf134</td>\n",
       "      <td>(RefSeq) enolase 4</td>\n",
       "      <td>10q25.3</td>\n",
       "      <td>hsa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Dmel_CG17654</td>\n",
       "      <td>Eno</td>\n",
       "      <td>(RefSeq) enolase, isoform B</td>\n",
       "      <td>2L</td>\n",
       "      <td>dme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>b2779</td>\n",
       "      <td>eno</td>\n",
       "      <td>(RefSeq) enolase</td>\n",
       "      <td>complement(2906643..2907941)</td>\n",
       "      <td>eco</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2645</td>\n",
       "      <td>GCK, FGQTL3, GK, GLK, HHF3, HK4, HKIV, HXKP, L...</td>\n",
       "      <td>(RefSeq) glucokinase</td>\n",
       "      <td>7p13</td>\n",
       "      <td>hsa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>b2388</td>\n",
       "      <td>glk</td>\n",
       "      <td>(RefSeq) glucokinase</td>\n",
       "      <td>complement(2508461..2509426)</td>\n",
       "      <td>eco</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                              short  \\\n",
       "0           5211                        PFKL, ATP-PFK, PFK-B, PFK-L   \n",
       "1           5213  PFKM, ATP-PFK, GSD7, PFK-1, PFK-A, PFK1, PFKA,...   \n",
       "2           5214                  PFKP, ATP-PFK, PFK-C, PFK-P, PFKF   \n",
       "3    Dmel_CG4001                                                Pfk   \n",
       "4          b1723                                               pfkB   \n",
       "..           ...                                                ...   \n",
       "56        387712                                    ENO4, C10orf134   \n",
       "57  Dmel_CG17654                                                Eno   \n",
       "58         b2779                                                eno   \n",
       "59          2645  GCK, FGQTL3, GK, GLK, HHF3, HK4, HKIV, HXKP, L...   \n",
       "60         b2388                                                glk   \n",
       "\n",
       "                                        long                      position  \\\n",
       "0   (RefSeq) phosphofructokinase, liver type                       21q22.3   \n",
       "1       (RefSeq) phosphofructokinase, muscle                      12q13.11   \n",
       "2     (RefSeq) phosphofructokinase, platelet                       10p15.2   \n",
       "3    (RefSeq) phosphofructokinase, isoform B                            2R   \n",
       "4          (RefSeq) 6-phosphofructokinase II              1806370..1807299   \n",
       "..                                       ...                           ...   \n",
       "56                        (RefSeq) enolase 4                       10q25.3   \n",
       "57               (RefSeq) enolase, isoform B                            2L   \n",
       "58                          (RefSeq) enolase  complement(2906643..2907941)   \n",
       "59                      (RefSeq) glucokinase                          7p13   \n",
       "60                      (RefSeq) glucokinase  complement(2508461..2509426)   \n",
       "\n",
       "   organism  \n",
       "0       hsa  \n",
       "1       hsa  \n",
       "2       hsa  \n",
       "3       dme  \n",
       "4       eco  \n",
       "..      ...  \n",
       "56      hsa  \n",
       "57      dme  \n",
       "58      eco  \n",
       "59      hsa  \n",
       "60      eco  \n",
       "\n",
       "[61 rows x 5 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM genes', conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Relational modeling\n",
    "\n",
    "For this part of the lab we would like to retrieve the enzyme information associated with each gene and the gene information associated with each pathway. Could this be accomplished another way than by adding additional fields to each table? Are the joins one-to-one, many-to-one, or one-to-many?\n",
    "\n",
    "In general, joins in biology are many-to-many, as for example due to the complex, multi-step nature of pathways, any given enzyme could be involved in many pathways, while at the same time pathways require many enzymes as cofactors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlite3.Cursor at 0x11cdac7a0>"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for tbl in ('enzyme_pathways', 'enzyme_genes'):\n",
    "    c.execute('DROP TABLE IF EXISTS %s;' % tbl);\n",
    "\n",
    "c.execute('''CREATE TABLE enzyme_pathways (\n",
    "    ec TEXT,\n",
    "    keggMap TEXT\n",
    ");''')\n",
    "c.executemany('INSERT INTO enzyme_pathways VALUES (?,?);', enz_path)\n",
    "\n",
    "c.execute('''CREATE TABLE enzyme_genes (\n",
    "    ec TEXT,\n",
    "    id TEXT\n",
    ");''')\n",
    "c.executemany('INSERT INTO enzyme_genes VALUES (?,?);', enz_genes)\n",
    "\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ec</th>\n",
       "      <th>keggMap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>map00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>map00030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6.2.1.5</td>\n",
       "      <td>map00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.8.1.4</td>\n",
       "      <td>map00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.8.1.4</td>\n",
       "      <td>map00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5.1.3.1</td>\n",
       "      <td>map00030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4.1.2.4</td>\n",
       "      <td>map00030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5.4.2.2</td>\n",
       "      <td>map00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5.4.2.2</td>\n",
       "      <td>map00030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4.2.1.3</td>\n",
       "      <td>map00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.2.4.1</td>\n",
       "      <td>map00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.2.4.1</td>\n",
       "      <td>map00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1.1.1.44</td>\n",
       "      <td>map00030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5.3.1.9</td>\n",
       "      <td>map00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5.3.1.9</td>\n",
       "      <td>map00030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4.2.1.11</td>\n",
       "      <td>map00010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2.7.1.2</td>\n",
       "      <td>map00010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ec   keggMap\n",
       "0   2.7.1.11  map00010\n",
       "1   2.7.1.11  map00030\n",
       "2    6.2.1.5  map00020\n",
       "3    1.8.1.4  map00010\n",
       "4    1.8.1.4  map00020\n",
       "5    5.1.3.1  map00030\n",
       "6    4.1.2.4  map00030\n",
       "7    5.4.2.2  map00010\n",
       "8    5.4.2.2  map00030\n",
       "9    4.2.1.3  map00020\n",
       "10   1.2.4.1  map00010\n",
       "11   1.2.4.1  map00020\n",
       "12  1.1.1.44  map00030\n",
       "13   5.3.1.9  map00010\n",
       "14   5.3.1.9  map00030\n",
       "15  4.2.1.11  map00010\n",
       "16   2.7.1.2  map00010"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM enzyme_pathways', conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ec</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>hsa:5211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>hsa:5213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>hsa:5214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>dme:Dmel_CG4001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.7.1.11</td>\n",
       "      <td>eco:b1723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>4.2.1.11</td>\n",
       "      <td>hsa:387712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>4.2.1.11</td>\n",
       "      <td>dme:Dmel_CG17654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>4.2.1.11</td>\n",
       "      <td>eco:b2779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>2.7.1.2</td>\n",
       "      <td>hsa:2645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>2.7.1.2</td>\n",
       "      <td>eco:b2388</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ec                id\n",
       "0   2.7.1.11          hsa:5211\n",
       "1   2.7.1.11          hsa:5213\n",
       "2   2.7.1.11          hsa:5214\n",
       "3   2.7.1.11   dme:Dmel_CG4001\n",
       "4   2.7.1.11         eco:b1723\n",
       "..       ...               ...\n",
       "56  4.2.1.11        hsa:387712\n",
       "57  4.2.1.11  dme:Dmel_CG17654\n",
       "58  4.2.1.11         eco:b2779\n",
       "59   2.7.1.2          hsa:2645\n",
       "60   2.7.1.2         eco:b2388\n",
       "\n",
       "[61 rows x 2 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM enzyme_genes', conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit ('base': conda)",
   "language": "python",
   "name": "python37364bitbasecondaf1e5bcae88af4e02a963c01f952d7725"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
